### PR TITLE
feat(core): improve type for `addInitScript()`

### PR DESCRIPTION
**Pull request by nix6839**
No callback parameter when no Arg.

### CODE REVIEW

**Comment by nix6839** [comment]
`// @ts-expect-error` comment only works with typescript 3.8+, so [this test](https://github.com/microsoft/playwright/actions/runs/3201696059/jobs/5229935120) fails. Can upgrade the typescript version? Or should I use another solution?
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments

File: utils/generate_types/test/test.ts [hunk 1]
```diff
@@ -586,6 +586,10 @@ playwright.chromium.launch().then(async browser => {
       }
     });
   }
+  {
+    // @ts-expect-error
+      await page.addInitScript((_expectNoArg: any) => {});
```
Comment (nix6839) [inline utils/generate_types/test/test.ts:591]:
```suggestion
    // @ts-expect-error
    await page.addInitScript((_expectNoArg: any) => {});
```

File: utils/generate_types/overrides.d.ts [hunk 2]
```diff
@@ -18,7 +18,12 @@ import { EventEmitter } from 'events';
 import { Readable } from 'stream';
 import { ReadStream } from 'fs';
 import { Protocol } from './protocol';
-import { Serializable, EvaluationArgument, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from './structs';
+import { Serializable, EvaluationArgument, PageFunction0, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from './structs';
+
+type AddInitScriptOptions = {
+  path?: string;
+  content?: string;
+}
```
Comment (nix6839) [inline utils/generate_types/overrides.d.ts:26]:
```suggestion
};
```

File: utils/generate_types/overrides.d.ts [hunk 3]
```diff
@@ -34,7 +39,8 @@ export interface Page {
   evaluateHandle<R, Arg>(pageFunction: PageFunction<Arg, R>, arg: Arg): Promise<SmartHandle<R>>;
   evaluateHandle<R>(pageFunction: PageFunction<void, R>, arg?: any): Promise<SmartHandle<R>>;
 
-  addInitScript<Arg>(script: PageFunction<Arg, any> | { path?: string, content?: string }, arg?: Arg): Promise<void>;
+  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>
+  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>
```
Comment (nix6839) [inline utils/generate_types/overrides.d.ts:43]:
```suggestion
  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>;
  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>;
```

File: utils/generate_types/overrides.d.ts [hunk 4]
```diff
@@ -100,7 +106,8 @@ export interface BrowserContext {
   exposeBinding(name: string, playwrightBinding: (source: BindingSource, arg: JSHandle) => any, options: { handle: true }): Promise<void>;
   exposeBinding(name: string, playwrightBinding: (source: BindingSource, ...args: any[]) => any, options?: { handle?: boolean }): Promise<void>;
 
-  addInitScript<Arg>(script: PageFunction<Arg, any> | { path?: string, content?: string }, arg?: Arg): Promise<void>;
+  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>
+  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>
```
Comment (nix6839) [inline utils/generate_types/overrides.d.ts:110]:
```suggestion
  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>;
  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>;
```

File: packages/playwright-core/types/types.d.ts [hunk 5]
```diff
@@ -19,7 +19,12 @@ import { EventEmitter } from 'events';
 import { Readable } from 'stream';
 import { ReadStream } from 'fs';
 import { Protocol } from './protocol';
-import { Serializable, EvaluationArgument, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from './structs';
+import { Serializable, EvaluationArgument, PageFunction0, PageFunction, PageFunctionOn, SmartHandle, ElementHandleForTag, BindingSource } from './structs';
+
+type AddInitScriptOptions = {
+  path?: string;
+  content?: string;
+}
```
Comment (nix6839) [inline packages/playwright-core/types/types.d.ts:27]:
```suggestion
};
```

File: packages/playwright-core/types/types.d.ts [hunk 6]
```diff
@@ -276,7 +281,35 @@ export interface Page {
...
+   * await page.addInitScript({ path: './preload.js' });
+   * ```
+   *
+   * > NOTE: The order of evaluation of multiple scripts installed via
+   * [browserContext.addInitScript(script[, arg])](https://playwright.dev/docs/api/class-browsercontext#browser-context-add-init-script)
+   * and [page.addInitScript(script[, arg])](https://playwright.dev/docs/api/class-page#page-add-init-script) is not defined.
+   * @param script Script to be evaluated in the page.
+   * @param arg Optional argument to pass to `script` (only supported when passing a function).
+   */
+  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>
```
Comment (nix6839) [inline packages/playwright-core/types/types.d.ts:312]:
```suggestion
  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>;
```

File: packages/playwright-core/types/types.d.ts [hunk 7]
```diff
@@ -276,7 +281,35 @@ export interface Page {
    * @param script Script to be evaluated in the page.
    * @param arg Optional argument to pass to `script` (only supported when passing a function).
    */
-  addInitScript<Arg>(script: PageFunction<Arg, any> | { path?: string, content?: string }, arg?: Arg): Promise<void>;
+  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>
```
Comment (nix6839) [inline packages/playwright-core/types/types.d.ts:284]:
```suggestion
  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>;
```

File: packages/playwright-core/types/types.d.ts [hunk 8]
```diff
@@ -6821,7 +6854,37 @@ export interface BrowserContext {
...
+   * });
+   * ```
+   *
+   * > NOTE: The order of evaluation of multiple scripts installed via
+   * [browserContext.addInitScript(script[, arg])](https://playwright.dev/docs/api/class-browsercontext#browser-context-add-init-script)
+   * and [page.addInitScript(script[, arg])](https://playwright.dev/docs/api/class-page#page-add-init-script) is not defined.
+   * @param script Script to be evaluated in all pages in the browser context.
+   * @param arg Optional argument to pass to `script` (only supported when passing a function).
+   */
+  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>
```
Comment (nix6839) [inline packages/playwright-core/types/types.d.ts:6887]:
```suggestion
  addInitScript(script: PageFunction0<any> | AddInitScriptOptions): Promise<void>;
```

File: packages/playwright-core/types/types.d.ts [hunk 9]
```diff
@@ -6821,7 +6854,37 @@ export interface BrowserContext {
    * @param script Script to be evaluated in all pages in the browser context.
    * @param arg Optional argument to pass to `script` (only supported when passing a function).
    */
-  addInitScript<Arg>(script: PageFunction<Arg, any> | { path?: string, content?: string }, arg?: Arg): Promise<void>;
+  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>
```
Comment (nix6839) [inline packages/playwright-core/types/types.d.ts:6857]:
```suggestion
  addInitScript<Arg>(script: PageFunction<Arg, any> | AddInitScriptOptions, arg: Arg): Promise<void>;
```

**Comment by pavelfeldman** [comment]
Could you file a bug explaining what does not work and what you are fixing? Thanks!

**Comment by pavelfeldman** [comment]
Closing as per request above.

**Pull request closed without merging**